language/move-vm/test-utils/src/gas_schedule.rs

```diff
@@ -17,88 +17,39 @@ use move_binary_format::{
+    gas_algebra::{
+        AbstractMemorySize, GasQuantity, InternalGas, InternalGasPerAbstractMemoryUnit,
+        InternalGasUnit, ToUnit, ToUnitFractional,
     },
     vm_status::StatusCode,
 };
 use move_vm_types::gas::GasMeter;
 use once_cell::sync::Lazy;
 use serde::{Deserialize, Serialize};
+use std::{
+    ops::{Add, Mul},
+    u64,
+};
 
-#[derive(Clone, Debug, Serialize, PartialEq, Deserialize)]
-pub struct GasConstants {
-    /// The cost per-byte read from global storage.
-    pub global_memory_per_byte_cost: InternalGasUnits<GasCarrier>,
-
-    /// The cost per-byte written to storage.
-    pub global_memory_per_byte_write_cost: InternalGasUnits<GasCarrier>,
-
-    /// The flat minimum amount of gas required for any transaction.
-    /// Charged at the start of execution.
-    pub min_transaction_gas_units: InternalGasUnits<GasCarrier>,
-
-    /// Any transaction over this size will be charged an additional amount per byte.
-    pub large_transaction_cutoff: AbstractMemorySize<GasCarrier>,
-
-    /// The units of gas that to be charged per byte over the `large_transaction_cutoff` in addition to
-    /// `min_transaction_gas_units` for transactions whose size exceeds `large_transaction_cutoff`.
-    pub intrinsic_gas_per_byte: InternalGasUnits<GasCarrier>,
-
-    /// ~5 microseconds should equal one unit of computational gas. We bound the maximum
-    /// computational time of any given transaction at roughly 20 seconds. We want this number and
-    /// `MAX_PRICE_PER_GAS_UNIT` to always satisfy the inequality that
-    /// MAXIMUM_NUMBER_OF_GAS_UNITS * MAX_PRICE_PER_GAS_UNIT < min(u64::MAX, GasUnits<GasCarrier>::MAX)
-    /// NB: The bound is set quite high since custom scripts aren't allowed except from predefined
-    /// and vetted senders.
-    pub maximum_number_of_gas_units: GasUnits<GasCarrier>,
-
-    /// The minimum gas price that a transaction can be submitted with.
-    pub min_price_per_gas_unit: GasPrice<GasCarrier>,
-
-    /// The maximum gas unit price that a transaction can be submitted with.
-    pub max_price_per_gas_unit: GasPrice<GasCarrier>,
+pub enum GasUnit {}
 
-    pub max_transaction_size_in_bytes: GasCarrier,
+pub type Gas = GasQuantity<GasUnit>;
 
-    pub gas_unit_scaling_factor: GasCarrier,
-    pub default_account_size: AbstractMemorySize<GasCarrier>,
+impl ToUnit<InternalGasUnit> for GasUnit {
+    const MULTIPLIER: u64 = 1000;
 }
 
-impl GasConstants {
-    pub fn to_internal_units(&self, units: GasUnits<GasCarrier>) -> InternalGasUnits<GasCarrier> {
-        InternalGasUnits::new(units.get() * self.gas_unit_scaling_factor)
-    }
-
-    pub fn to_external_units(&self, units: InternalGasUnits<GasCarrier>) -> GasUnits<GasCarrier> {
-        GasUnits::new(units.get() / self.gas_unit_scaling_factor)
-    }
+impl ToUnitFractional<GasUnit> for InternalGasUnit {
+    const NOMINATOR: u64 = 1;
+    const DENOMINATOR: u64 = 1000;
```

Comment:
A note on `GasStatus`: the existence of two gas units (internal and external) is a legacy of the old days. My plan is to replace it with a test-only gas meter impl that doesn't have this differentiation, which we can use in the Move repo. This gas meter will likely NOT be suitable for production use.

One thing that's a bit tricky to deal with is the Move CLI,  since it still uses `GasStatus` internally and expects clients to pass in an old `CostTable`. IMO we'll need to redesign this part and see if we can come up with a generic approach that works for everyone, perhaps by asking clients to pass in their own gas meter implementation.

I know the repeated simplification and eventual removal of `GasStatus` will be a problem for clients who are currently using it in prod (e.g. @jolestar) or planning to do so , but IMO the correct thing to do really is to implement your own gas meter. In the worst case, you can fork the old `GasStatus` and try to make it work.

Reply:
Do you have a plan to eliminate the `gas_unit_scaling_factor` and `InternalGasUnits`? 

I think the `gas_unit_scaling_factor` is designed for Diem's Currency that does not have enough decimals and introduce the internal gas units to fix the bug https://github.com/diem/diem/issues/7238. 

We set the `gas_unit_scaling_factor` to 1 on Starcoin, so InternalGasUnits == GasUnits.

I support eliminating the `gas_unit_scaling_factor` and `InternalGasUnits`.

Reply:
> Do you have a plan to eliminate the gas_unit_scaling_factor and InternalGasUnits?

I would say the plan is to remove `GasStatus` so we won't have an official gas meter implementation inside the Move repo any more. Instead, with the changes I recently made, each blockchain can and should implement its own gas meter, in a way that suits its own needs. Whether to have the `gas_unit_scaling_factor` and two gas units is now a design choice you can make on your own. Again, I highly encourage you to replace `GasStatus` with your own gas meter soon.

The new gas meter implementation I mentioned was only meant for testing within the Move repo, where there is no need to calculate any sort of transaction fee, and thus no need to have two gas units to deal with precision issues.

With that being said I think I may end up renaming `InternalGasUnits` to something like `VMGasUnits`. The name `InternalGasUnits` seems to imply there exists some external gas units, which isn't necessarily true. In the mid to long term, it's possible that the concept of `InternalGasUnits` may disappear from the Move repo, but only after I'm done with the VM ownership refactoring and if my plan to further generify gas metering works out.

> I think the gas_unit_scaling_factor is designed for Diem's Currency that does not have enough decimals and introduce the internal gas units to fix the bug https://github.com/diem/diem/issues/7238.

As Tim pointed out, one advantage of having a scaled gas unit is to allow one to "change the scaling factor as time goes on as, e.g., the instructions become less expensive computationally, whereas once a given precision is picked for a currency that can't really ever be changed. And in general, the scaling factor gives us a knob to turn to scale the transaction fees in the system as time goes on". 

We at Aptos also considered whether we should unify the two gas units and eventually decided to keep them, for the reason Tim brought up, but I guess you could live without scaling if you have enough decimals.

Reply:
Discussed offline.